shell/platform/fuchsia/dart_runner/dart_test_component_controller_v2.cc

```diff
@@ -0,0 +1,647 @@
+    fidl::InterfaceRequest<fuchsia::test::CaseIterator> request,
+    async_dispatcher_t* dispatcher,
+    fit::function<void(CaseIterator*)> done_callback)
+    : binding_(this, std::move(request), dispatcher),
+      done_callback_(std::move(done_callback)) {}
+
+// |fuchsia::test::CaseIterator|
+static const char kTestCaseName[] = "dart_test_v2";
+void DartTestComponentControllerV2::CaseIterator::GetNext(
+    GetNextCallback callback) {
+  if (get_next_call_count == 0) {
+    fuchsia::test::Case test_case;
+    test_case.set_name(std::string(kTestCaseName));
+    test_case.set_enabled(true);
+    std::vector<fuchsia::test::Case> cases;
+    cases.push_back(std::move(test_case));
+    callback(std::move(cases));
+    get_next_call_count++;
+
+  } else {
+    std::vector<fuchsia::test::Case> cases;
+    callback(std::move(cases));
+    // this would be removed
+    done_callback_(this);
+  }
+}
+
+// |fuchsia::test::CaseIterator|
+std::unique_ptr<DartTestComponentControllerV2::CaseIterator>
+DartTestComponentControllerV2::RemoveCaseInterator(
+    CaseIterator* case_iterator) {
+  auto it = case_iterators_.find(case_iterator);
+  std::unique_ptr<DartTestComponentControllerV2::CaseIterator>
+      case_iterator_ptr;
+  if (it != case_iterators_.end()) {
+    case_iterator_ptr = std::move(it->second);
+    case_iterators_.erase(it);
+  }
+  return case_iterator_ptr;
+}
+
+// |fuchsia::test::Suite|
+void DartTestComponentControllerV2::GetTests(
+    fidl::InterfaceRequest<fuchsia::test::CaseIterator> iterator) {
+  auto case_iterator =
+      std::make_unique<CaseIterator>(std::move(iterator), loop_->dispatcher(),
+                                     [this](CaseIterator* case_iterator) {
+                                       RemoveCaseInterator(case_iterator);
+                                     });
+  case_iterators_.emplace(case_iterator.get(), std::move(case_iterator));
+}
+
+// |fuchsia::test::Suite|
+void DartTestComponentControllerV2::Run(
+    std::vector<fuchsia::test::Invocation> tests,
+    fuchsia::test::RunOptions options,
+    fidl::InterfaceHandle<fuchsia::test::RunListener> listener) {
+  RunDartMain();
+
+  std::vector<std::string> args;
+  if (options.has_arguments()) {
+    args = std::move(*options.mutable_arguments());
+  }
+
+  if (options.has_parallel()) {
+    FML_LOG(WARNING) << "Ignoring 'parallel'. Pass test specific flags, eg: "
+                        "for rust test pass in "
+                        "--test-threads="
+                     << options.parallel();
+  }
+
+  auto listener_proxy = listener.Bind();
+
+  for (auto it = tests.begin(); it != tests.end(); it++) {
+    auto invocation = std::move(*it);
+    std::string test_case_name;
+    if (invocation.has_name()) {
+      test_case_name = invocation.name();
+    }
+
+    zx::socket out, err, out_client, err_client;
+    auto status = zx::socket::create(0, &out, &out_client);
+    if (status != ZX_OK) {
+      FML_LOG(FATAL) << "cannot create socket: "
```

Comment:
To distinguish these error messages, can we make the out one "cannot create out socket:" and the error one "cannot create error socket:"?